src/utils/number.ts

```diff
@@ -1,4 +1,98 @@
 import { numberKeys } from "./key";
+import { createLocaleNumberFormatter, getDecimalSeparator, getMinusSign } from "./locale";
+
+// regex for number sanitization
+const allLeadingZerosOptionallyNegative = /^([-0])0+(?=\d)/;
+const decimalAtEndOfStringButNotStart = /(?!^\.)\.$/;
+const allHyphensExceptTheStart = /(?!^-)-/g;
+const isNegativeDecimalOnlyZeros = /^-\b0\b\.?0*$/;
+
+// adopted from https://stackoverflow.com/a/66939244
+export class BigDecimal {
+  value: bigint;
+
+  // BigInt("-0").toString() === "0" which removes the minus sign when typing numbers like -0.1
+  isNegative: boolean;
+
+  // Configuration: constants
+  static DECIMALS = 100; // number of decimals on all instances
+
+  static ROUNDED = true; // numbers are truncated (false) or rounded (true)
+
+  static SHIFT = BigInt("1" + "0".repeat(BigDecimal.DECIMALS)); // derived constant
+
+  constructor(input: string | BigDecimal) {
+    if (input instanceof BigDecimal) {
+      return input;
+    }
+    const [ints, decis] = String(input).split(".").concat("");
+    this.value =
+      BigInt(ints + decis.padEnd(BigDecimal.DECIMALS, "0").slice(0, BigDecimal.DECIMALS)) +
+      BigInt(BigDecimal.ROUNDED && decis[BigDecimal.DECIMALS] >= "5");
+
+    this.isNegative = input.charAt(0) === "-";
+  }
+
+  static _divRound(dividend: bigint, divisor: bigint): bigint {
+    return BigDecimal.fromBigInt(
+      dividend / divisor + (BigDecimal.ROUNDED ? ((dividend * BigInt(2)) / divisor) % BigInt(2) : BigInt(0))
+    );
+  }
+
+  static fromBigInt(bigint: bigint): bigint {
+    return Object.assign(Object.create(BigDecimal.prototype), { value: bigint });
+  }
+
+  toString(): string {
+    const s = this.value
+      .toString()
+      .replace(new RegExp("-", "g"), "")
+      .padStart(BigDecimal.DECIMALS + 1, "0");
+
+    const i = s.slice(0, -BigDecimal.DECIMALS);
+    const d = s.slice(-BigDecimal.DECIMALS).replace(/\.?0+$/, "");
+    const value = i.concat(d.length ? "." + d : "");
+    return (this.isNegative ? "-" : "").concat(value);
+  }
+
+  formatToParts(locale: string, numberingSystem?: string): Intl.NumberFormatPart[] {
+    const formatter = createLocaleNumberFormatter(locale, numberingSystem);
+
+    const s = this.value
+      .toString()
+      .replace(new RegExp("-", "g"), "")
+      .padStart(BigDecimal.DECIMALS + 1, "0");
+
+    const i = s.slice(0, -BigDecimal.DECIMALS);
+    const d = s.slice(-BigDecimal.DECIMALS).replace(/\.?0+$/, "");
+
+    const parts = formatter.formatToParts(BigInt(i));
```

Comment:
I'm wondering if this line ☝️ could read like this, that way the next line could be removed:
```
const parts = formatter.formatToParts(BigInt(this.isNegative ? "-" + i : i));
```

Reply:
I tried that, it would cause the minus sign to be removed when typing negative franctional numbers like `-0.1`. There is no such thing as `-0` for `BigInt` like there is for `number`. https://forum.kirupa.com/t/js-tip-of-the-day-negative-zero/643094